Windows: complete rewrite of drive restoration and image writing

grulja opened the pull request:
This is a complete rewrite of the Windows support. It introduces a new library 'libwindisk' that combines use of WMI and WinAPI to manipulate with devices. We use Windows Storage Management for getting information about devices and some actions, especially since pure WinAPI does not seem to have support for things like formatting or partition removal. This library is used in both the app and the helper process used for writing and formatting. We now also don't rely on diskpart since it didn't work properly in some cases.
    
Fixes #626 | Fixes #575 | Fixes #574 |Fixes #555 | Fixes #96

gastoner commented on src/helper/win/writejob.cpp:
```diff
@@ -20,397 +21,422 @@
+            writtenBytes = drive.write(static_cast<char *>(outBuffer), readBytes);
+            if (writtenBytes < 0) {
+                m_err << tr("Failed to write to the device: ") << drive.errorString() << "\n";
+                m_err << tr("Failed to write to the device:") << getLastError() << "\n";
```
```suggestion
                m_err << tr("Failed to write to the device: ") << getLastError() << "\n";
```

gastoner commented on src/helper/win/writejob.cpp:
```diff
@@ -20,397 +21,422 @@
+            writtenBytes = drive.write(static_cast<char *>(outBuffer), sectorSize);
+            if (writtenBytes < 0) {
+                m_err << tr("Failed to write to the device: ") << drive.errorString() << "\n";
+                m_err << tr("Failed to write to the device:") << getLastError() << "\n";
```
```suggestion
                m_err << tr("Failed to write to the device: ") << getLastError() << "\n";
```

gastoner commented on src/helper/win/writejob.cpp:
```diff
@@ -20,397 +21,422 @@
+        writtenBytes = drive.write(static_cast<char *>(buffer), readBytes);
+        if (writtenBytes < 0) {
+            m_err << tr("Failed to write to the device: ") << drive.errorString() << "\n";
+            m_err << tr("Failed to write to the device:") << getLastError() << "\n";
```
```suggestion
            m_err << tr("Failed to write to the device: ") << getLastError() << "\n";
```

gastoner reviewed (commented):
It looks like for mine only USB drive it does not want to write the iso image

![image](https://github.com/user-attachments/assets/d707f083-96eb-4798-998f-d418dc9050b9)

But the restore worked without any problems

grulja commented:
> > Hi @kparal, do you think you can do some testing? You reported some of the mentioned issues in this PR and your testing was always super helpful.
> 
> I'll try to find the time to look at it. Do you intend to release it before F41 is out? (If we manage to approve it this week, it would be Tuesday next week). If you intend to keep it in testing for a few weeks or so, then I'd know I don't need to rush so much.

I would like to get it released before F41, because it should avoid all the issues users were experiencing before on Windows, but it assumes the code I wrote does what it's supposed to.

frantisekz commented:
> Hi @kparal, do you think you can do some testing? You reported some of the mentioned issues in this PR and your testing was always super helpful.
> 
> You can download Windows installer with this change from the latest CI run [here](https://github.com/FedoraQt/MediaWriter/actions/runs/11403098275).

It is borked on a clean Windows 10 installation due to missing MSVCP140.dll (hard crash on launch).

grulja commented:
> > Hi @kparal, do you think you can do some testing? You reported some of the mentioned issues in this PR and your testing was always super helpful.
> > You can download Windows installer with this change from the latest CI run [here](https://github.com/FedoraQt/MediaWriter/actions/runs/11403098275).
> 
> It is borked on a clean Windows 10 installation due to missing MSVCP140.dll (hard crash on launch).

You need https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170, but that was needed even before.

frantisekz commented:
> You need https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170, but that was needed even before.

In that case not a regression, but it'd be better from UX if that could show the link/info or even handle the installation of the dependency (I mean, if it's possible in the Windows world :) )?

kparal commented:
I downloaded the [latest build](https://github.com/FedoraQt/MediaWriter/actions/runs/11478794389).

My experiences from Windows 11:
1. I was mildly confused by the [erase confirmation dialog](https://i.imgur.com/UQZVanS.png). I don't understand how the size of the stick (my stick is 64GB, btw) is relevant to the question whether I want to overwrite it or not.
2. The very first time I wrote an image, I was hit by the 4.8% verification bug (#669). I also noticed that at that very time, the File Explorer showed new disks D: and E: as present. Every other time I wrote an image, no new disks were presented, and the later verification was not hit by the 4.8% bug.
3. I still can't write the image twice in a row, without disconnecting it and reconnecting it first (#574).
4. After reboot, the first image write failed with [this error](https://i.imgur.com/dpH431F.png). But where can I grab logs? It would be nice if FMW told me. (Eventually, I found them in ~/Documents, but it would be nice to tell the user, if they want to report a bug. Also, missing timestamps make it hard to figure out whether they are relevant). [Log1](https://pastebin.com/raw/rkGUwEkH), [Log2](https://pastebin.com/raw/q57caNfc). This looks repeatable, happened again after another reboot.
5. Further image writes work OK.
6. When I finish writing and disconnect the disk, the main screen doesn't update and still shows "Restore $diskname" as one of the options.
7. I also tested USB HDD, and worked OK, same as the flash drives.

grulja commented:
> I downloaded the [latest build](https://github.com/FedoraQt/MediaWriter/actions/runs/11478794389).
> 
> My experiences from Windows 11:
> 
> 1. I was mildly confused by the [erase confirmation dialog](https://i.imgur.com/UQZVanS.png). I don't understand how the size of the stick (my stick is 64GB, btw) is relevant to the question whether I want to overwrite it or not.

Yeah, might not have been the best idea. Basically, this tries to ensure you don't accidentally wipe out your external hard rive. I might probably check for bigger sizes, like  > 64GB, which would be more close to real hard drive size.

> 2. The very first time I wrote an image, I was hit by the 4.8% verification bug ([Media Writer produces USB stick with FAIL media check #669](https://github.com/FedoraQt/MediaWriter/issues/669)). I also noticed that at that very time, the File Explorer showed new disks D: and E: as present. Every other time I wrote an image, no new disks were presented, and the later verification was not hit by the 4.8% bug.

Interesting, looks we now have a reproducer for this 4,8% verification bug.

> 3. I still can't write the image twice in a row, without disconnecting it and reconnecting it first ([Can't write a Live image twice: Destination drive is not writable #574](https://github.com/FedoraQt/MediaWriter/issues/574)).

What error did you get in this case? 

> 4. After reboot, the first image write failed with [this error](https://i.imgur.com/dpH431F.png). But where can I grab logs? It would be nice if FMW told me. (Eventually, I found them in ~/Documents, but it would be nice to tell the user, if they want to report a bug. Also, missing timestamps make it hard to figure out whether they are relevant). [Log1](https://pastebin.com/raw/rkGUwEkH), [Log2](https://pastebin.com/raw/q57caNfc). This looks repeatable, happened again after another reboot.

This looks like the same issue @gastoner is experiencing. I haven't been able to reproduce it myself, but I will try today to install Windows 11 on yet another laptop to try to reproduce this. It might be that the device driver is diffent in VM.

> 5. Further image writes work OK.
> 6. When I finish writing and disconnect the disk, the main screen doesn't update and still shows "Restore $diskname" as one of the options.
> 7. I also tested USB HDD, and worked OK, same as the flash drives.

The rest was without any error even after a reboot? I wonder what is so special about the 64GB stick you used.

kparal commented:
> Yeah, might not have been the best idea. Basically, this tries to ensure you don't accidentally wipe out your external hard rive. I might probably check for bigger sizes, like > 64GB, which would be more close to real hard drive size.

My recommendation is to drop the heuristic and just ask every time. Include the real size of the medium in the dialog, e.g. like this:
_The whole device (all of XYZ GB) will be erased and the selected image will be written onto it. Do you want to continue?_

> Interesting, looks we now have a reproducer for this 4,8% verification bug.

No, unfortunately not. I saw it again a few times today, but I'm still not sure how to trigger it. It seems like some kind of a race. Might be even in Windows itself.

> I still can't write the image twice in a row, without disconnecting it and reconnecting it first

[Screenshot](https://i.imgur.com/PHHa2iA.png) [Log1](https://pastebin.com/raw/Ayzfif1N) [Log2](https://pastebin.com/raw/T7JX07gR)

> After reboot, the first image write failed

Sigh, it's not happening today. What a mess.

> I wonder what is so special about the 64GB stick you used.

That's not what I meant. I don't think my 64GB stick is special. I just wanted to say that USB HDDs also work in general (same as USB sticks work in general, but fail in certain cases).

kparal commented:
To conclude my testing, it seems to work to a similar extent as the old code. Generally it's fine, but in some cases it fails for me. You probably hoped for a better answer, sorry :-)

grulja commented:
Hi @gastoner @kparal, can you try latest build?
1) Is not fixed. It's unrelated to this Windows rewrite.
2) No idea, but it is probably caused by Windows mounting the drive automatically and creating "System Volume Information" on the drive and that makes the media check to fail. I might have done a change that avoids it, but not sure.
3+) Hopefully fixed or at least improved

Also please try restoring the device.

kparal commented:
Well, this is interesting. I tested about 6 writes, both to a thumb drive and an external hdd, and **all of them** fail with the 4.8% bug. The write phase proceeds fine, but the verification phase is always almost immediately skipped (I see only the first segment of the progress bar filled out, and then it immediately it skips to Finished!), and then it always fails verification (`checkisomd5`).

So whatever you did, it somehow made the check phase be skipped (possibly it checks 4.8% and then skips the rest), and the 4.8% bug now happens every time.

grulja commented:
> Well, this is interesting. I tested about 6 writes, both to a thumb drive and an external hdd, and **all of them** fail with the 4.8% bug. The write phase proceeds fine, but the verification phase is always almost immediately skipped (I see only the first segment of the progress bar filled out, and then it immediately it skips to Finished!), and then it always fails verification (`checkisomd5`).
> 
> So whatever you did, it somehow made the check phase be skipped (possibly it checks 4.8% and then skips the rest), and the 4.8% bug now happens every time.

Alright, the `checkisomd5` is fixed. I used Qt before, then changed it to use directly WinAPI to write and for some reason it fails later with the verifacation. Going back to Qt fixes it. Still, @gastoner reported before that the writing is still broken for him. I was not able to install Windows 11 on my another laptop, but Windows 11 in VM and Windows 10 on an old laptop do work reliably. I'm running out of ideas what could possibly be wrong.

grulja commented:
@kparal @gastoner can you please do one more round of testing?

I have installed Windows 11 and could reproduce the error. I have changed few things and it made things more reliable. Still I got verification error afterwards, but not all the time. This was only with faster USB drive so I suspect it was caused by some data still not being fully written. I have added a `flush()` call after a successful write to ensure the last data is written in hope it fixes the problem for good. I did one try with that and it worked.

Thank you.

kparal commented:
I tried it about 6 times. I haven't seen the 4.8% bug at all. I haven't seen the double-write bug. I still saw the first-write-after-boot bug ([screenshot](https://i.imgur.com/1PaQ4XH.png) [log1](https://pastebin.com/raw/yFJzB4Qr) [log2](https://pastebin.com/raw/qhrnCjq3)) , but only with my thumb drive, not usb hdd. Also, waiting a few minutes after rebooting seemed to avoid this bug.

gastoner commented:
Here are mine observations: 
1. When I try to format drive with Rufus (nonbootable - GPT or MBR) - FMW will write the iso successfuly ✅ 
2. When I try to write the iso aggain (without restoring using FMW) - it failes [FedoraMediaWriter-helper.log](https://github.com/user-attachments/files/17534894/FedoraMediaWriter-helper.log) ❌ 
3. When I try to write the iso (using Rufus) then restoring using FMW and writing iso using FMW - was successful ✅ 
4. After restoring (using FMW) next writing was not successful (logs are the same) ❌ 

With using Rufus to write the iso, from log I see some similar patterns as in FMW-helper.log [rufus.log](https://github.com/user-attachments/files/17535010/rufus.log) 
`Like Could not get Disk Extents: [0x00000001] Incorrect function.` or `Write error [0x000001B1] A device which does not exist was specified.`

grulja commented:
Hi @kparal, can you do one more (last) round of testing? @gastoner says this version finally works for him too. I will then merge this and make a new release of FMW.

grulja commented:
I'm going to merge this now. I will make something like a pre-release so this gets more tested and we can catch the remaining corner cases before making a "stable" release.

kparal commented:
I've spent several hours (sigh) testing 5.1.90 pre-release with two different usb sticks, and it mostly works fine. I only filed #759 as a regression.

However, I still see some problems, randomly and rarely. I saw the 4.8% problem once, and I saw the Restore functionality only re-creating a partition but not formatting it several times. But I'm starting to attribute it to some problem in Windows itself. I noticed that every time I have a Fedora image on the stick and plug it in, the "Safely remove" systray icon only contains "..." when I display its menu, instead of "Eject $thumbdrive". It's like some operation is pending and never finishes. It's even worse in the Disk Management tool, if you try to eject the drive from there, the app completely freezes. When I play with the system like this, often I then saw the inability to restore the drive, and I had to reboot the PC to make it work again. Some operation was stuck/pending/etc in the background, probably, blocking access. But not always. It's exhausting, because when I think I finally have a reproducer, I find out that it works differently with a different stick. It's probably some race in Windows itself.

So overall 5.1.90 seems to work OK, and the issues I randomly see might just be Windows bugs, which are not going to disappear with any FMW rewrite...

grulja commented:
> I've spent several hours (sigh) testing 5.1.90 pre-release with two different usb sticks, and it mostly works fine. I only filed #759 as a regression.
> 
> However, I still see some problems, randomly and rarely. I saw the 4.8% problem once, and I saw the Restore functionality only re-creating a partition but not formatting it several times. But I'm starting to attribute it to some problem in Windows itself. I noticed that every time I have a Fedora image on the stick and plug it in, the "Safely remove" systray icon only contains "..." when I display its menu, instead of "Eject $thumbdrive". It's like some operation is pending and never finishes. It's even worse in the Disk Management tool, if you try to eject the drive from there, the app completely freezes. When I play with the system like this, often I then saw the inability to restore the drive, and I had to reboot the PC to make it work again. Some operation was stuck/pending/etc in the background, probably, blocking access. But not always. It's exhausting, because when I think I finally have a reproducer, I find out that it works differently with a different stick. It's probably some race in Windows itself.
> 
> So overall 5.1.90 seems to work OK, and the issues I randomly see might just be Windows bugs, which are not going to disappear with any FMW rewrite...

Thank you so much for a thorough testing. I really appreciate it. Do you think in a normal use, where the user just writes the image, boot it and restores the drive aftewards this will work reliably? Basically asking whether the issues you mentioned are only reproducible when doing multiple operations one after the other? I will look into the issue you reported, I already have some ideas what could be wrong.

grulja commented:
@kparal would you be able to get a log (MediaWriter-helper.txt) from when the restore functionality fails to format the partition?

kparal commented:
Here we go:
[FedoraMediaWriter-helper.log](https://github.com/user-attachments/files/17735006/FedoraMediaWriter-helper.log)

But I couldn't reproduce it with just the systray interaction this time, I had to open Disk Management, try to eject the drive from there, kill the app, reinsert the thumb drive, and then restore in FMW.

On an unrelated note, I noticed that when the drive is affected by th 4.8% bug, it shows up in explorer (D:, "unformatted") and in systray correctly (drive name). When it's not affected by 4.8%, it doesn't show up in explorer and systray only contains "..." instead of drive name. But I can't be sure, this is hard to reproduce several times in a row.

kparal commented:
> Do you think in a normal use, where the user just writes the image, boot it and restores the drive aftewards this will work reliably? Basically asking whether the issues you mentioned are only reproducible when doing multiple operations one after the other? 

The fewer operations, the higher likelihood of things working, I assume. But it's tricky. I might have a Windows computer where I create the thumb drive, then use it to install Fedora on a different computer, and then put the thumb drive back to the Windows computer to restore it. That's already two actions. And it might be different if you closed FMW in the meantime or kept it running. And I'm unable to pinpoint a simple reproducer to my issues, so I can't say how much reliable a simple use case is. But overall I think this is definitely not worse than previous FMW versions, if that helps :-) In other words, I think you can release it this way.

kparal commented:
[This](https://github.com/FedoraQt/MediaWriter/issues/575#issuecomment-2474122671) could be a big source of my troubles. Maybe if we fix that, all of these issues might go away.
